File: pakalon-cli/src/providers/health.ts
/**
 * Provider Health Check Utilities for Pakalon CLI
 *
 * Monitors health and latency of local and cloud providers.
 */

import logger from "@/utils/logger.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ProviderType = "local" | "cloud";
export type HealthStatus = "healthy" | "degraded" | "down";

export interface ProviderHealth {
  providerId: string;
  providerType: ProviderType;
  status: HealthStatus;
  latencyMs: number;
  lastChecked: Date;
  error?: string;
  modelCount?: number;
}

export interface HealthCheckOptions {
  timeout?: number;
  retries?: number;
  retryDelay?: number;
}

// ---------------------------------------------------------------------------
// Default Configurations
// ---------------------------------------------------------------------------

const DEFAULT_OPTIONS: HealthCheckOptions = {
  timeout: 5000,
  retries: 2,
  retryDelay: 1000,
};

// ---------------------------------------------------------------------------
// Provider Health Checker
// ---------------------------------------------------------------------------

export class ProviderHealthChecker {
  private healthCache: Map<string, ProviderHealth> = new Map();
  private lastCheck: Map<string, number> = new Map();

  /**
   * Check health of a single provider
   */
  async checkProvider(
    providerId: string,
    baseUrl: string,
    providerType: ProviderType,
    options: HealthCheckOptions = {}
  ): Promise<ProviderHealth> {
    const opts = { ...DEFAULT_OPTIONS, ...options };
    const startTime = Date.now();

    try {
      if (providerType === "local") {
        await this.checkLocalProvider(baseUrl, opts);
      } else {
        await this.checkCloudProvider(baseUrl, opts);
      }

      const latency = Date.now() - startTime;

      const health: ProviderHealth = {
        providerId,
        providerType,
        status: latency < 1000 ? "healthy" : "degraded",
        latencyMs: latency,
        lastChecked: new Date(),
      };

      this.healthCache.set(providerId, health);
      this.lastCheck.set(providerId, Date.now());
      return health;
    } catch (error) {
      const latency = Date.now() - startTime;
      const health: ProviderHealth = {
        providerId,
        providerType,
        status: "down",
        latencyMs: latency,
        lastChecked: new Date(),
        error: error instanceof Error ? error.message : String(error),
      };

      this.healthCache.set(providerId, health);
      this.lastCheck.set(providerId, Date.now());
      return health;
    }
  }

  /**
   * Check local provider (Ollama or LM Studio)
   */
  private async checkLocalProvider(baseUrl: string, options: HealthCheckOptions): Promise<void> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), options.timeout);

    try {
      // Try Ollama endpoint first
      try {
        const response = await fetch(`${baseUrl}/api/tags`, {
          method: "GET",
          signal: controller.signal,
        });
        if (response.ok) {
          return;
        }
      } catch {
        // Ignore Ollama error, try LM Studio
      }

      // Try LM Studio endpoint
      try {
        const response = await fetch(`${baseUrl}/v1/models`, {
          method: "GET",
          signal: controller.signal,
        });
        if (response.ok) {
          return;
        }
      } catch {
        // Ignore LM Studio error
      }

      throw new Error(`Cannot connect to provider at ${baseUrl}`);
    } finally {
      clearTimeout(timeout);
    }
  }

  /**
   * Check cloud provider (OpenRouter)
   */
  private async checkCloudProvider(baseUrl: string, options: HealthCheckOptions): Promise<void> {
    const apiKey = process.env.OPENROUTER_API_KEY;
    if (!apiKey) {
      throw new Error("OpenRouter API key not configured");
    }

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), options.timeout);

    try {
      const response = await fetch(`${baseUrl}/models`, {
        method: "GET",
        headers: {
          Authorization: `Bearer ${apiKey}`,
        },
        signal: controller.signal,
      });

      if (!response.ok) {
        throw new Error(`OpenRouter returned status ${response.status}`);
      }
    } finally {
      clearTimeout(timeout);
    }
  }

  /**
   * Get cached health status for a provider
   */
  getHealth(providerId: string): ProviderHealth | undefined {
    return this.healthCache.get(providerId);
  }

  /**
   * Get health status for all providers
   */
  getAllHealth(): ProviderHealth[] {
    return Array.from(this.healthCache.values());
  }

  /**
   * Check health of all configured providers
   */
  async checkAllProviders(): Promise<ProviderHealth[]> {
    const results: ProviderHealth[] = [];

    // Check local providers
    const ollamaUrl = process.env.PAKALON_OLLAMA_URL || "http://localhost:11434";
    const lmstudioUrl = process.env.PAKALON_LMSTUDIO_URL || "http://localhost:1234";

    const ollamaHealth = await this.checkProvider("ollama", ollamaUrl, "local");
    results.push(ollamaHealth);

    const lmstudioHealth = await this.checkProvider("lmstudio", lmstudioUrl, "local");
    results.push(lmstudioHealth);

    // Check cloud providers
    if (process.env.OPENROUTER_API_KEY) {
      const openrouterHealth = await this.checkProvider(
        "openrouter",
        "https://openrouter.ai/api/v1",
        "cloud"
      );
      results.push(openrouterHealth);
    }

    return results;
  }

  /**
   * Check if a provider is healthy
   */
  isProviderHealthy(providerId: string): boolean {
    const health = this.healthCache.get(providerId);
    return health !== undefined && health.status !== "down";
  }

  /**
   * Get list of healthy provider IDs
   */
  getHealthyProviders(): string[] {
    return Array.from(this.healthCache.entries())
      .filter(([_, health]) => health.status !== "down")
      .map(([id, _]) => id);
  }

  /**
   * Clear health cache
   */
  clearCache(): void {
    this.healthCache.clear();
    this.lastCheck.clear();
  }
}

// Global instance
export const healthChecker = new ProviderHealthChecker();

// ---------------------------------------------------------------------------
// Convenience Functions
// ---------------------------------------------------------------------------

/**
 * Check provider health
 */
export async function checkProviderHealth(
  providerId: string,
  baseUrl: string,
  providerType: ProviderType,
  options?: HealthCheckOptions
): Promise<ProviderHealth> {
  return healthChecker.checkProvider(providerId, baseUrl, providerType, options);
}

/**
 * Check all providers health
 */
export async function checkAllProvidersHealth(): Promise<ProviderHealth[]> {
  return healthChecker.checkAllProviders();
}

/**
 * Get provider health status
 */
export function getProviderHealth(providerId: string): ProviderHealth | undefined {
  return healthChecker.getHealth(providerId);
}

/**
 * Check if provider is healthy
 */
export function isProviderHealthy(providerId: string): boolean {
  return healthChecker.isProviderHealthy(providerId);
}

File: pakalon-backend/app/health.py
"""Provider health check system for Pakalon backend.

Monitors health and latency of local and cloud providers.
"""

import logging
import time
from datetime import datetime
from typing import Any, Literal

import httpx
from pydantic import BaseModel

from app.config import get_settings
from app.features import get_feature_flags

logger = logging.getLogger(__name__)


class ProviderHealth(BaseModel):
    """Health status of a provider."""

    provider_id: str
    provider_type: Literal["local", "cloud"]
    status: Literal["healthy", "degraded", "down"]
    latency_ms: float
    last_checked: datetime
    error: str | None = None
    model_count: int = 0


class ProviderHealthChecker:
    """Checks health of all configured providers."""

    def __init__(self):
        self._health_cache: dict[str, ProviderHealth] = {}
        self._last_check: dict[str, float] = {}

    async def check_provider(
        self, provider_id: str, base_url: str, provider_type: Literal["local", "cloud"]
    ) -> ProviderHealth:
        """Check health of a single provider."""
        start_time = time.time()

        try:
            if provider_type == "local":
                await self._check_local_provider(base_url)
            else:
                await self._check_cloud_provider(base_url)

            latency = (time.time() - start_time) * 1000  # Convert to ms

            health = ProviderHealth(
                provider_id=provider_id,
                provider_type=provider_type,
                status="healthy" if latency < 1000 else "degraded",
                latency_ms=latency,
                last_checked=datetime.now(),
            )
        except Exception as e:
            latency = (time.time() - start_time) * 1000
            health = ProviderHealth(
                provider_id=provider_id,
                provider_type=provider_type,
                status="down",
                latency_ms=latency,
                last_checked=datetime.now(),
                error=str(e),
            )

        self._health_cache[provider_id] = health
        self._last_check[provider_id] = time.time()
        return health

    async def _check_local_provider(self, base_url: str) -> None:
        """Check local provider (Ollama or LM Studio)."""
        async with httpx.AsyncClient(timeout=5.0) as client:
            # Try Ollama endpoint first
            try:
                response = await client.get(f"{base_url.rstrip('/')}/api/tags")
                if response.status_code == 200:
                    return
            except httpx.RequestError:
                pass

            # Try LM Studio endpoint
            try:
                response = await client.get(f"{base_url.rstrip('/')}/v1/models")
                if response.status_code == 200:
                    return
            except httpx.RequestError:
                pass

            raise ConnectionError(f"Cannot connect to provider at {base_url}")

    async def _check_cloud_provider(self, base_url: str) -> None:
        """Check cloud provider (OpenRouter)."""
        settings = get_settings()

        if not settings.openrouter_master_key:
            raise ValueError("OpenRouter API key not configured")

        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                f"{base_url.rstrip('/')}/models",
                headers={"Authorization": f"Bearer {settings.openrouter_master_key}"},
            )
            if response.status_code != 200:
                raise ConnectionError(f"OpenRouter returned status {response.status_code}")

    def get_health(self, provider_id: str) -> ProviderHealth | None:
        """Get cached health status for a provider."""
        return self._health_cache.get(provider_id)

    def get_all_health(self) -> list[ProviderHealth]:
        """Get health status for all providers."""
        return list(self._health_cache.values())

    async def check_all_providers(self) -> list[ProviderHealth]:
        """Check health of all configured providers."""
        settings = get_settings()
        flags = get_feature_flags()
        results = []

        # Check local providers
        if flags.local_models:
            if settings.local_ollama_enabled:
                health = await self.check_provider(
                    "ollama", settings.local_ollama_url, "local"
                )
                results.append(health)

            if settings.local_lmstudio_enabled:
                health = await self.check_provider(
                    "lmstudio", settings.local_lmstudio_url, "local"
                )
                results.append(health)

        # Check cloud providers
        if flags.cloud_providers and settings.openrouter_master_key:
            health = await self.check_provider(
                "openrouter", "https://openrouter.ai/api/v1", "cloud"
            )
            results.append(health)

        return results

    def is_provider_healthy(self, provider_id: str) -> bool:
        """Check if a provider is healthy based on cached status."""
        health = self._health_cache.get(provider_id)
        return health is not None and health.status != "down"

    def get_healthy_providers(self) -> list[str]:
        """Get list of healthy provider IDs."""
        return [
            pid
            for pid, health in self._health_cache.items()
            if health.status != "down"
        ]


# Global instance
health_checker = ProviderHealthChecker()
